pkg/ccl/backupccl/backup_processor.go

```diff
@@ -709,7 +710,34 @@ func (s *sstSink) push(ctx context.Context, resp returnedSST) error {
 	s.queueSize += len(resp.sst)
 
 	if s.queueSize >= int(s.queueCap) {
-		sort.Slice(s.queue, func(i, j int) bool { return s.queue[i].f.Span.Key.Compare(s.queue[j].f.Span.Key) < 0 })
+		sort.Slice(s.queue, func(i, j int) bool {
+			cmp := s.queue[i].f.Span.Key.Compare(s.queue[j].f.Span.Key)
+			// If two backup files have the same start key, it implies we have
+			// paginated in the middle of a key's revisions and those revisions
+			// straddle the file boundary. We must ensure that these files are sorted
+			// in decreasing order of the key's revision timestamps so that they are
+			// added in-order to the underlying SSTWriter.
+			if cmp == 0 {
+				// Since the start key of the files are equal we compare the EndKeys.
+				cmp := s.queue[i].f.Span.EndKey.Compare(s.queue[j].f.Span.EndKey)
+				// If the end keys are also equal, we know that both files only contain
+				// revisions of the same key, and so it is safe to sort in descending
+				// order of the timestamp of the last revision in each file.
+				if cmp == 0 {
+					return s.queue[j].endKeyTS.Less(s.queue[i].endKeyTS)
```

Comment:
Confirming: This is correct even if the case where the file ends at a key boundary and endKeyTS is empty because that empty timestamp will sort last which is where it should be anyway.

Reply:
Yup, added a subtest to the new unit test for this case.